nss/aad_auth/error_c.go

```diff
@@ -0,0 +1,49 @@
+package main
+
+/*
+#include <errno.h>
+#include <nss.h>
+
+typedef enum nss_status nss_status;
+*/
+import "C"
+import (
+	"context"
+	"errors"
+
+	"github.com/ubuntu/aad-auth/internal/logger"
+	"github.com/ubuntu/aad-auth/internal/nss"
+)
+
+// errToCStatus converts our Go errors to corresponding nss status returned code and errno.
+// If err is nil, it returns a success.
+func errToCStatus(ctx context.Context, err error) (nssStatus, errno int) {
+	nssStatus = C.NSS_STATUS_SUCCESS
+
+	switch {
+	case errors.Is(err, nss.ErrTryAgainEAgain):
+		nssStatus = C.NSS_STATUS_TRYAGAIN
+		errno = C.EAGAIN
+	case errors.Is(err, nss.ErrTryAgainERange):
+		nssStatus = C.NSS_STATUS_TRYAGAIN
+		errno = C.ERANGE
+	case errors.Is(err, nss.ErrUnavailableENoEnt):
+		nssStatus = C.NSS_STATUS_UNAVAIL
+		errno = C.ENOENT
+	case errors.Is(err, nss.ErrNotFoundENoEnt):
+		nssStatus = C.NSS_STATUS_NOTFOUND
+		errno = C.ENOENT
+	case errors.Is(err, nss.ErrNotFoundSuccess):
+		nssStatus = C.NSS_STATUS_SUCCESS
+		errno = C.ENOENT
+	case err != nil: // Unexpected returned error
+		nssStatus = C.NSS_STATUS_SUCCESS
+		errno = C.EINVAL
+	}
+
+	if err != nil {
+		logger.Debug(ctx, "%v", err)
+	}
+
```

Comment:
We don't have the errnop* anymore. You commented this code yesterday and we did got over it real quick, so I thought there was no need for it anymore.

Reply:
We are still converting to errno and nssStatus still, no? I’m not sure to understand you here.

Reply:
We are, but now we return them directly in the function, rather than storing the C value in a pointer. I can just add a logger call to Debug both values, but they are covered indirectly in the fmtOutput tests

Reply:
The idea is to not cover them in tests, but just to print something to the end user in syslogs for debugging purposes! So yes, printing both if err != nil will be of interest